### ISSUE TITLE
MRTK V2.1 - Toggle does not set dimension to 1 at first call

### ISSUE THREAD

**Issue by arsi0001** [microsoft/MixedRealityToolkit-Unity#6745]
I updated to MRTK v2.1 and right now I'm struggling with the toggles from my [previous post][1]. Don't get me wrong, the issue in the linked post got fixed! But now I'm having a different problem.<br><br>
**What happens:**<br>
I have a settings panel in my scene and the component `DebugContent.cs` attached to it. Every time I open it, `OnEnable` triggers and sets the state of my toggles according to third parameter of `SetToggleState()`. So lets say that the third parameter is always true → `DiagnosticsSystem.ShowProfiler`, `debugProvider.enabled` and `speechInputHandler.enabled` are true!<br><br> On MRTK v2.0 I had no issues, but now on v2.1 my toggles are not getting toggled by the first time. **I have to open my panel a second time so the toggles are toggled**!<br>

    //OnEnable in DebugContent.cs
    private void OnEnable()
    {
        // Set Profiler, Log viewer, Speech Input
        Utils.SetToggleState(profilerTgl, profilerTxt, DiagnosticsSystem.ShowProfiler);
        Utils.SetToggleState(logViewerTgl, logViewTxt, debugProvider.enabled);
        Utils.SetToggleState(speechInputTgl, speechInputTxt, speechInputHandler.enabled);
    }

    //SetToggleState in Utils.cs
    public static void SetToggleState(Interactable comp, TextMesh tm, bool isOn)
    {
        Debug.Log("Util >> SetToggleState >> Set toggle → " + comp.gameObject.name +" to → " + isOn + " >> Comp is enable → " + comp.enabled + " and has dimension → " + comp.CurrentDimension);
        if (isOn)
        {
            comp.CurrentDimension = 1;
            tm.text = "On";
        }
        else
        {
            comp.CurrentDimension = 0;
            tm.text = "Off";
        }
        Debug.Log("Util >> SetToggleState >> Toggle set to dimension → " + comp.CurrentDimension);
    }
**More on:**<br>
The strange part is that the first time I open my panel the text of my toggles shows `On` but my toggle is not toggled. Only if I open my panel the second time the toggle get toggled:
[![enter image description here][2]][2]<br>
The output of my `Debug.Log` is also interesting, because the current dimension is every time 0:<br>
*First time:*<br>
`Util >> SetToggleState >> Set toggle → Debug_Tgl to → True >> Comp is enabled → True and has dimension → 0` <br>`Util >> SetToggleState >> Toggle set to dimension → 1`<br>
*Second time:*<br>
`Util >> SetToggleState >> Set toggle → Debug_Tgl to → True >> Comp is enabled → True and has dimension → 0`<br> `Util >> SetToggleState >> Toggle set to dimension → 1`

I'm using `CurrentDimension`, because `SetDimensionIndex` is now deprecated.

  [1]: https://github.com/microsoft/MixedRealityToolkit-Unity/issues/6241#issue-504291216
  [2]: https://i.stack.imgur.com/YujlN.png

## Your Setup

- Unity Version 2019
- MRTK Version v2.1

## Target Platform

- HoloLens

## [Stackoverflow](https://stackoverflow.com/questions/59067678/mrtk-v2-1-toggle-does-not-set-dimension-to-1-at-first-call)

**Comment by julenka** [comment]
Catherine may already have a fix for this actually coming out in 2.2.

**Comment by arsi0001** [comment]
ok, thanks for answering. Do you know a workaround? Or should I just trigger twice?

**Comment by julenka** [comment]
Friendly ping to @CDiaz-MS  to answer arsi0001 question

**Comment by CDiaz-MS** [comment]
Currently out of office and I have not investigated this issue, but the first thing that comes to mind is to try using  comp.IsToggled = true/false instead of comp.CurrentDimension to set the toggle states within Interactable.

**Comment by Troy-Ferrell** [comment]
I was able to repro this in 2.2. The issue is because of a race condition between the parent script OnEnable and Interactable.Awake().

In the toggle awake(), we set the CurrentDimension to the desired startDimensionIndex set in the inspector. This is what overrides the users' setting in their SetToggleState

@arsi0001 to workarond this on your side. In your DebugContent.cs, add a trigger in your Start(). This worked in my test. 

@julenka / @CDiaz-MS , not exactly sure though how we could "fix" this on MRTK side. We can't execute code earlier than Awake. A bit hacky, but we could add a bool in CurrentDimension if the setter has ever been triggered before. Then in on Awake, don't set the startDimensionIndex if someone has already overrided the CurrentDimension. 

```
// DebugContent.cs
private void Start()
{
SetToggles();
}
private void OnEnable()
{
SetToggles();
}

private void SetToggles()
{
    // Set Profiler, Log viewer, Speech Input
    Utils.SetToggleState(profilerTgl, profilerTxt, DiagnosticsSystem.ShowProfiler);
    Utils.SetToggleState(logViewerTgl, logViewTxt, debugProvider.enabled);
    Utils.SetToggleState(speechInputTgl, speechInputTxt, speechInputHandler.enabled);
}
```
